File: src/types/Track.ts
import { FullUser, User, UserModel } from './User'

// The general case models come from the "/v1" routes
export type TrackArtwork = {
  ['150x150']: string | null
  ['480x480']: string | null
  ['1000x1000']: string
}

export type RemixOf<T> = {
  tracks: T[]
}

export type RemixParent = {
  parent_track_id: string
}

export type Track = {
  artwork: TrackArtwork,
  description: string
  downloadable: boolean
  duration: number
  favorite_count: number
  genre: string
  id: string
  mood: string
  permalink: string
  play_count: number
  release_date: string
  remix_of: RemixOf<RemixParent> | null
  repost_count: number
  tags: string | null
  title: string
  user: User
}

export type TrackDownload = {
  cid: string | null
  is_downloadable: boolean
  requires_follow: boolean
}

export type TrackFieldVisibility = {
  genre: boolean
  mood: boolean
  play_count: boolean
  remixes: boolean
  share: boolean
  tags: boolean
}

export type Repost = {
  repost_item_id: string
  repost_type: string
  user_id: string
}

export type Favorite = {
  favorite_item_id: string
  favorite_type: string
  user_id: string
}

export type StemParent = {
  category: string | null
  parent_track_id: number | null
}

export type TrackSegment = {
  duration: number
  multihash: string
}

// Prefixed with "full", these models come from the "v1/full" route
export type FullRemixParent = {
  has_remix_author_reposted: boolean
  has_remix_author_saved: boolean
  parent_track_id: string
  user: FullUser
}

export type FullTrack = Track & {
  blocknumber: number
  cover_art_sizes: string
  cover_art: string
  create_date: string
  created_at: string
  credits_splits: string
  download: TrackDownload
  field_visibility: TrackFieldVisibility
  followee_favorites: Favorite[]
  followee_reposts: Repost[]
  has_current_user_reposted: boolean
  has_current_user_saved: boolean
  is_delete: boolean
  is_unlisted: boolean
  isrc: string
  iswc: string
  license: string
  remix_of: RemixOf<FullRemixParent> | null
  route_id: string
  stem_of: StemParent
  track_segments: TrackSegment[]
  updated_at: string
  user_id: string
  user: FullUser
}

// Suffixed with "model", these types come from the non-v1 API routes
export type RemixParentModel = {
  has_remix_author_reposted: boolean
  has_remix_author_saved: boolean
  parent_track_id: number
  user: UserModel
}

export type TrackModel = {
  blockhash: string
  blocknumber: number
  cover_art_sizes: string
  cover_art: string
  create_date: string
  created_at: string
  credits_splits: string
  description: string
  download: TrackDownload
  field_visibility: TrackFieldVisibility
  file_type: string
  followee_reposts: Repost[]
  followee_saves: Favorite[]
  has_current_user_reposted: boolean
  has_current_user_saved: boolean
  genre: string
  is_current: boolean
  is_delete: boolean
  is_unlisted: boolean
  isrc: string
  iswc: string
  length: number
  license: string
  metadata_multihash: string
  mood: string
  owner_id: number
  permalink: string
  play_count: number
  release_date: string
  remix_of: RemixOf<RemixParentModel> | null
  repost_count: number
  route_id: string
  save_count: number
  stem_of: null
  tags: string | null
  title: string
  track_id: number
  track_segments: TrackSegment[]
  txhash: string
  updated_at: string
  user: UserModel
}

File: discovery-provider/src/models/tracks/track.py
from sqlalchemy import (
    Boolean,
    Column,
    DateTime,
    ForeignKey,
    Integer,
    PrimaryKeyConstraint,
    String,
    Text,
    text,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import relationship, validates
from src.model_validator import ModelValidator
from src.models.base import Base
from src.models.model_utils import (
    RepresentableMixin,
    get_fields_to_validate,
    validate_field_helper,
)
from src.models.tracks.track_route import TrackRoute
from src.models.users.user import User


class Track(Base, RepresentableMixin):
    __tablename__ = "tracks"

    blockhash = Column(ForeignKey("blocks.blockhash"))  # type: ignore
    track_id = Column(Integer, primary_key=True, nullable=False)
    is_current = Column(Boolean, primary_key=True, nullable=False)
    is_delete = Column(Boolean, nullable=False)
    owner_id = Column(Integer, nullable=False, index=True)
    title = Column(Text)
    length = Column(Integer)
    cover_art = Column(String)
    tags = Column(String)
    genre = Column(String)
    mood = Column(String)
    credits_splits = Column(String)
    create_date = Column(String)
    release_date = Column(String)
    file_type = Column(String)
    metadata_multihash = Column(String)
    blocknumber = Column(ForeignKey("blocks.number"), index=True)  # type: ignore
    track_segments = Column(JSONB(), nullable=False)
    created_at = Column(DateTime, nullable=False, index=True)
    description = Column(String)
    isrc = Column(String)
    iswc = Column(String)
    license = Column(String)
    updated_at = Column(DateTime, nullable=False)
    cover_art_sizes = Column(String)
    download = Column(JSONB())
    is_unlisted = Column(Boolean, nullable=False, server_default=text("false"))
    field_visibility = Column(JSONB())
    route_id = Column(String)
    stem_of = Column(JSONB())
    remix_of = Column(JSONB())
    txhash = Column(
        String,
        primary_key=True,
        nullable=False,
        server_default=text("''::character varying"),
    )
    slot = Column(Integer)
    is_available = Column(Boolean, nullable=False, server_default=text("true"))

    block = relationship(  # type: ignore
        "Block", primaryjoin="Track.blockhash == Block.blockhash"
    )
    block1 = relationship(  # type: ignore
        "Block", primaryjoin="Track.blocknumber == Block.number"
    )

    _routes = relationship(  # type: ignore
        TrackRoute,
        primaryjoin="and_(\
            remote(Track.track_id) == foreign(TrackRoute.track_id),\
            TrackRoute.is_current)",
        lazy="joined",
        viewonly=True,
    )

    user = relationship(  # type: ignore
        User,
        primaryjoin="and_(\
            remote(Track.owner_id) == foreign(User.user_id),\
            User.is_current)",
        lazy="joined",
        viewonly=True,
    )

    @property
    def _slug(self):
        return self._routes[0].slug if self._routes else ""

    @property
    def permalink(self):
        if self.user and self.user[0].handle and self._slug:
            return f"/{self.user[0].handle}/{self._slug}"
        return ""

    PrimaryKeyConstraint(is_current, track_id, txhash)

    ModelValidator.init_model_schemas("Track")
    fields = get_fields_to_validate("Track")

    # unpacking args into @validates
    @validates(*fields)
    def validate_field(self, field, value):
        return validate_field_helper(field, value, "Track", getattr(Track, field).type)
